pkg/ipamd/ipamd.go

```diff
@@ -738,31 +737,37 @@ func (c *IPAMContext) tryAssignIPs() (increasedPool bool, err error) {
 // 2) set up linux ENI related networking stack.
 // 3) add all ENI's secondary IP addresses to datastore
 func (c *IPAMContext) setupENI(eni string, eniMetadata awsutils.ENIMetadata, trunkENI string) error {
+	primaryENI := c.awsClient.GetPrimaryENI()
 	// Add the ENI to the datastore
-	err := c.dataStore.AddENI(eni, eniMetadata.DeviceNumber, eni == c.awsClient.GetPrimaryENI(), eni == trunkENI)
+	err := c.dataStore.AddENI(eni, eniMetadata.DeviceNumber, eni == primaryENI, eni == trunkENI)
 	if err != nil && err.Error() != datastore.DuplicatedENIError {
 		return errors.Wrapf(err, "failed to add ENI %s to data store", eni)
 	}
+	// Store the primary IP of the ENI
+	c.primaryIP[eni] = eniMetadata.PrimaryIPv4Address()
 
 	// For secondary ENIs, set up the network
-	if eni != c.awsClient.GetPrimaryENI() {
-		eniPrimaryIP := eniMetadata.PrimaryIPv4Address()
-		err = c.networkClient.SetupENINetwork(eniPrimaryIP, eniMetadata.MAC, eniMetadata.DeviceNumber, eniMetadata.SubnetIPv4CIDR)
+	if eni != primaryENI {
+		err = c.networkClient.SetupENINetwork(c.primaryIP[eni], eniMetadata.MAC, eniMetadata.DeviceNumber, eniMetadata.SubnetIPv4CIDR)
 		if err != nil {
+			// Failed to set up the ENI
+			errRemove := c.dataStore.RemoveENIFromDataStore(eni, true)
+			if errRemove != nil {
+				log.Warnf("failed to remove ENI %s: %v", eni, errRemove)
+			}
+			delete(c.primaryIP, eni)
 			return errors.Wrapf(err, "failed to set up ENI %s network", eni)
 		}
 	}
 
-	c.primaryIP[eni] = c.addENIaddressesToDataStore(eniMetadata.IPv4Addresses, eni)
+	c.addENIaddressesToDataStore(eniMetadata.IPv4Addresses, eni)
```

Comment:
Should we handle the case of eniMetadata not being in sync here?

Reply:
We do call `setupENI()` from `tryAllocateENI()`, and since #1174 we should not add any empty ENIs to the datastore. If we only get half the secondary IPs, we will add those, and then if IMDS gets updated, we will check that with EC2 later in the reconcile loop.